SSA/Projects/Scf/Scf.lean

```diff
@@ -0,0 +1,175 @@
+  Expr.mk
+    (op := .iterate k)
+    (ty_eq := rfl)
+    (args := .cons input .nil)
+    (regArgs := HVector.cons body HVector.nil)
+
+def if_ {Γ : Ctxt _} {t : Ty} (cond : Var Γ .bool) (v : Var Γ t) (then_ else_ : Com Op [t] t) : Expr Op Γ t :=
+  Expr.mk
+    (op := .if t)
+    (ty_eq := rfl)
+    (args := .cons cond <| .cons v .nil)
+    (regArgs := HVector.cons then_ <| HVector.cons else_ <| HVector.nil)
+
+def run {Γ : Ctxt _} {t : Ty} (v : Var Γ t) (body : Com Op [t] t) : Expr Op Γ t :=
+  Expr.mk
+    (op := .run t)
+    (ty_eq := rfl)
+    (args := .cons v .nil)
+    (regArgs := HVector.cons body <| HVector.nil)
+
+def for_ {Γ : Ctxt _} {t : Ty} (niter : Var Γ .nat) (v : Var Γ t) (body : Com Op [.nat, t] t) : Expr Op Γ t :=
+  Expr.mk
+    (op := .for t)
+    (ty_eq := rfl)
+    (args := .cons niter <| .cons v .nil)
+    (regArgs := HVector.cons body <| HVector.nil)
+
+/-- 'if' condition of a true variable evaluates to the true region body. -/
+theorem if_true {t : Ty} (cond : Var Γ .bool) (hcond : Γv cond = true) (v : Var Γ t) (then_ else_ : Com Op [t] t) :
+  Expr.denote (ScfRegion.if_ (t := t) cond v then_ else_) Γv =
+  Expr.denote (ScfRegion.run (t := t) v then_) Γv := by
+    simp[Expr.denote, if_, run]
+    simp_peephole[hcond] at Γv
+    simp[ite_true]
+
+/-- 'if' condition of a true variable evaluates to the true region body. -/
+theorem if_false {t : Ty} (cond : Var Γ .bool) (hcond : Γv cond = false) (v : Var Γ t) (then_ else_ : Com Op [t] t) :
+  Expr.denote (ScfRegion.if_ (t := t) cond v then_ else_) Γv =
+  Expr.denote (ScfRegion.run (t := t) v else_) Γv := by
+    simp[Expr.denote, if_, run]
+    simp_peephole[hcond] at Γv
+    simp[ite_true]
+
+attribute [local simp] Ctxt.snoc
+
+/-- running `f(x) = x + x` 0 times is the identity. -/
+def lhs : Com Op [.int] .int :=
+  Com.lete (iterate (k := 0) ⟨0, by simp[Ctxt.snoc]⟩ (
+      Com.lete (add ⟨0, by simp[Ctxt.snoc]⟩ ⟨0, by simp[Ctxt.snoc]⟩) -- fun x => (x + x)
+      <| Com.ret ⟨0, by simp[Ctxt.snoc]⟩
+  )) <|
+  Com.ret ⟨0, by simp[Ctxt.snoc]⟩
+
+def rhs : Com Op [.int] .int :=
+  Com.ret ⟨0, by simp[Ctxt.snoc]⟩
+
+attribute [local simp] Ctxt.snoc
+
+set_option pp.proofs false in
+set_option pp.proofs.withType false in
+noncomputable def p1 : PeepholeRewrite Op [.int] .int:=
+  { lhs := lhs, rhs := rhs, correct := by
+      rw [lhs, rhs]
+      funext Γv
+      /-
+      Com.denote
+        (Com.lete
+          (iterate 0 { val := 0, property := lhs.proof_1 }
+            (Com.lete (add { val := 0, property := lhs.proof_1 } { val := 0, property := lhs.proof_1 })
+              (Com.ret { val := 0, property := lhs.proof_2 })))
+          (Com.ret { val := 0, property := lhs.proof_2 }))
+        Γv =
+      Com.denote (Com.ret { val := 0, property := rhs.proof_1 }) Γv
+      -/
+      simp_peephole [add, iterate] at Γv
+      /-  ∀ (a : BitVec 32), (fun v => v + v)^[0] a = a -/
+      simp [Function.iterate_zero]
+      done
+  }
```

Comment:
Can we define a similar peephole rewrite for the "if true" and "if false" rewrites?

Reply:
We could also hardcode the region, and later replace it with a mvar.

Reply:
What about this comment?